### PR TITLE
Add and edit Rank, Unit, Company and Platoon fields

### LINKED ISSUE #38: Add contact fields relevant to military personnel
As an admin clerk working with military personnel, I can keep track of the military personnel's rank, unit, company and platoon.

Sample format of `add` command (only rank is required):
`add n/NAME p/PHONE_NUMBER e/EMAIL a/ADDRESS [t/TAG]…​ r/RANK [u/UNIT] [c/COMPANY] [pl/PLATOON]`

### LINKED ISSUE #47: Edit contact fields relevant to military personnel
i.e rank, unit, company, and platoon

### PR DESCRIPTION
**Pull request by jugheadjones10**
closes #38
closes #47

### CODE REVIEW

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 1]
```diff
@@ -25,39 +25,42 @@ public class TypicalPersons {
...
+            .withEmail("heinz@example.com").withAddress("wall street").withRank("CPL").build();
     public static final Person DANIEL = new PersonBuilder().withName("Daniel Meier").withPhone("87652533")
-            .withEmail("cornelia@example.com").withAddress("10th street").withTags("friends").build();
+            .withEmail("cornelia@example.com").withAddress("10th street").withRank("3SG").withTags("friends").build();
     public static final Person ELLE = new PersonBuilder().withName("Elle Meyer").withPhone("9482224")
-            .withEmail("werner@example.com").withAddress("michegan ave").build();
+            .withEmail("werner@example.com").withAddress("michegan ave").withRank("2LT").build();
     public static final Person FIONA = new PersonBuilder().withName("Fiona Kunz").withPhone("9482427")
-            .withEmail("lydia@example.com").withAddress("little tokyo").build();
+            .withEmail("lydia@example.com").withAddress("little tokyo").withRank("private").build();
```
Comment (rexcyrio) [inline src/test/java/seedu/address/testutil/TypicalPersons.java:42]: I think `"private"` should be `"PTE"`

File: src/test/java/seedu/address/testutil/PersonBuilder.java [hunk 2]
```diff
@@ -20,11 +24,20 @@ public class PersonBuilder {
     public static final String DEFAULT_PHONE = "85355255";
     public static final String DEFAULT_EMAIL = "amy@gmail.com";
     public static final String DEFAULT_ADDRESS = "123, Jurong West Ave 6, #08-111";
+    public static final String DEFAULT_RANK = "private";
```
Comment (rexcyrio) [inline src/test/java/seedu/address/testutil/PersonBuilder.java:27]: I think `"private"` should be `"PTE"`

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 3]
```diff
@@ -25,39 +25,42 @@ public class TypicalPersons {
...
     public static final Person GEORGE = new PersonBuilder().withName("George Best").withPhone("9482442")
-            .withEmail("anna@example.com").withAddress("4th street").build();
+            .withEmail("anna@example.com").withAddress("4th street").withRank("CPL").build();
 
     // Manually added
     public static final Person HOON = new PersonBuilder().withName("Hoon Meier").withPhone("8482424")
-            .withEmail("stefan@example.com").withAddress("little india").build();
+            .withEmail("stefan@example.com").withAddress("little india").withRank("private").build();
```
Comment (rexcyrio) [inline src/test/java/seedu/address/testutil/TypicalPersons.java:48]: I think `"private"` should be `"PTE"`

**Review by niekis — CHANGES_REQUESTED** [review]
I think you might be missing some test cases, especially there should probably be one under test/java/seedu/address/model/person/RankTest
maybe can just look through the [tutorial 2](https://nus-cs2103-ay2223s2.github.io/tp/tutorials/AddRemark.html) to see if you're missing any other tests

**Comment by minosx31** [comment]
> Speaking of test cases, I also noted that some of the Gradle test cases failed. Are those test cases still relelvant? Should the code / test cases be modified such that all the test cases pass?

Yeah, I think need to take a look at the test cases and see if can modify to fit the change

**Comment by jugheadjones10** [comment]
@rexcyrio @minosx31 @niekis 
Fixed all tests broken by the new fields, Gradle is building now. Regarding field-specific tests, maybe we can add those in future iterations?

File: src/main/resources/view/PersonListCard.fxml [hunk 4]
```diff
@@ -1,36 +1,34 @@
...
-        <Label fx:id="name" text="\$first" styleClass="cell_big_label" />
-      </HBox>
-      <FlowPane fx:id="tags" />
-      <Label fx:id="phone" styleClass="cell_small_label" text="\$phone" />
-      <Label fx:id="address" styleClass="cell_small_label" text="\$address" />
-      <Label fx:id="email" styleClass="cell_small_label" text="\$email" />
-    </VBox>
-  </GridPane>
+<?import javafx.scene.control.*?>
+<?import javafx.scene.layout.*?>
```
Comment (rexcyrio) [inline src/main/resources/view/PersonListCard.fxml:5]: I think we should use explicit imports here.

File: src/test/java/seedu/address/logic/parser/AddressBookParserTest.java [hunk 5]
```diff
@@ -49,7 +49,7 @@ public void parseCommand_clear() throws Exception {
     @Test
     public void parseCommand_delete() throws Exception {
         DeleteCommand command = (DeleteCommand) parser.parseCommand(
-                DeleteCommand.COMMAND_WORD + " " + INDEX_FIRST_PERSON.getOneBased());
+            DeleteCommand.COMMAND_WORD + " " + INDEX_FIRST_PERSON.getOneBased());
```
Comment (rexcyrio) [inline src/test/java/seedu/address/logic/parser/AddressBookParserTest.java:52]: Remember to set the continuation indent in `Settings > Editor > Code Style > Java > Tabs and Indents` to 8 spaces! (see [the CS2103T Java coding standard](https://se-education.org/guides/conventions/java/intermediate.html#:~:text=Indentation%20for%20wrapped%20lines%20should%20be%208%20spaces))

**Review by rexcyrio — CHANGES_REQUESTED** [review]

**Comment by rexcyrio** [comment]
> @rexcyrio @minosx31 @niekis Fixed all tests broken by the new fields, Gradle is building now. Regarding field-specific tests, maybe we can add those in future iterations?

I think we should add the test cases for `Rank` now rather than in future iterations.

It doesn't really make sense to "accept" a new feature only to introduce test cases for the feature later on -- are we saying that we accepted a potentially buggy feature previously?

Maybe one good test case to have is to check whether we accept `"pte"` (i.e., lowercase characters instead of uppercase characters).

**Comment by jugheadjones10** [comment]
> > @rexcyrio @minosx31 @niekis Fixed all tests broken by the new fields, Gradle is building now. Regarding field-specific tests, maybe we can add those in future iterations?
> 
> I think we should add the test cases for `Rank` now rather than in future iterations.
> 
> It doesn't really make sense to "accept" a new feature only to introduce test cases for the feature later on -- are we saying that we accepted a potentially buggy feature previously?
> 
> Maybe one good test case to have is to check whether we accept `"pte"` (i.e., lowercase characters instead of uppercase characters).

True. I'll add the tests.

File: src/test/java/seedu/address/model/person/RankTest.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Test
+    public void constructor_null_throwsNullPointerException() {
+        assertThrows(NullPointerException.class, () -> new Rank(null));
+    }
+
+    @Test
+    public void constructor_invalidRank_throwsIllegalArgumentException() {
+        String invalidRank = "";
+        assertThrows(IllegalArgumentException.class, () -> new Email(invalidRank));
```
Comment (rexcyrio) [inline src/test/java/seedu/address/model/person/RankTest.java:19]: I think you meant to test `Rank(invalidRank)` here instead of `Email(invalidRank)`

**Review by rexcyrio — CHANGES_REQUESTED** [review]
I think you accidentally committed the `/bin` folder to the repo. 

To fix this:

1. delete the `/bin` folder and commit
1. add the following line to `.gitignore` and commit

```gitignore
# VS Code Red Hat Java build files
/bin/
```

([Referenced from StackOverflow](https://stackoverflow.com/questions/28388597/removing-bin-executable-folder-from-github))

You can also consider disabling the "autobuild" functionality of Red Hat Java so that it'll stop creating the `/bin` folder for you automatically.

Code_I4vr9lQnz6

**Pull request merged**